src/planner/binder/query_node/bind_setop_node.cpp

```diff
@@ -182,6 +184,57 @@ static void BuildUnionByNameInfo(ClientContext &context, BoundSetOperationNode &
 	}
 }
 
+void Binder::BindCollationGroup(unique_ptr<BoundSetOperationNode> &bound_set_op) {
+	if (bound_set_op->left->type != QueryNodeType::SELECT_NODE ||
+	    bound_set_op->right->type != QueryNodeType::SELECT_NODE) {
+		return;
+	}
+	auto &left_node = bound_set_op->left->Cast<BoundSelectNode>();
+	auto &left_bind_state = left_node.bind_state;
+	auto &right_node = bound_set_op->right->Cast<BoundSelectNode>();
+	auto &right_bind_state = right_node.bind_state;
+
+	// using set data structure to ensure uniqueness
+	std::set<idx_t> collation_indexes(left_node.collation_sel_idx.begin(), left_node.collation_sel_idx.end());
+	std::copy(right_node.collation_sel_idx.begin(), right_node.collation_sel_idx.end(),
+	          std::inserter(collation_indexes, collation_indexes.end()));
+
+	// verifies collation conflicts
+	for (idx_t collate_idx : collation_indexes) {
+		auto &left_expr = left_bind_state.original_expressions[collate_idx];
+		auto &right_expr = right_bind_state.original_expressions[collate_idx];
+		// at least one expression must have to be a collation
+		D_ASSERT(left_expr->GetExpressionClass() == ExpressionClass::COLLATE ||
```

Comment:
Collations are stored in types - collate expressions are only one way to get collations into the system. The `default_collation` can apply a collation to all string columns, and individual columns in a table can have a `COLLATE` attribute. As such, collations need to be handled when the types are propagated through the set operation.